extensions/azurecore/src/account-provider/utils/msalFileEncryptionHelper.ts

```diff
@@ -0,0 +1,52 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+import * as azdata from 'azdata';
+import * as crypto from 'crypto';
+import { AuthLibrary } from '../../constants';
+import { FileEncryptionHelper } from './fileEncryptionHelper';
+
+/**
+ * Designed to work with MSAL.NET cache using the same 'aes-256-cbc' algorithm.
+ * Key and IV are shared through credential store.
+ */
+export class MSALFileEncryptionHelper extends FileEncryptionHelper {
+	constructor(_credentialService: azdata.CredentialProvider, _fileName: string) {
+		super(_credentialService, _fileName);
+		this._authLibrary = AuthLibrary.MSAL;
+	}
+
+	private _ivBuffer: Buffer | undefined;
+	private _keyBuffer: Buffer | undefined;
+
+	async init(): Promise<void> {
+		const iv = await this.readEncryptionKey(`${this._fileName}-iv`);
+		const key = await this.readEncryptionKey(`${this._fileName}-key`);
+		if (!iv || !key) {
+			this._ivBuffer = crypto.randomBytes(16);
+			this._keyBuffer = crypto.randomBytes(32);
+			await this.saveEncryptionKey(`${this._fileName}-iv`, this._ivBuffer.toString('utf16le'));
+			await this.saveEncryptionKey(`${this._fileName}-key`, this._keyBuffer.toString('utf16le'));
+		} else {
+			this._ivBuffer = Buffer.from(iv, 'utf16le');
+			this._keyBuffer = Buffer.from(key, 'utf16le');
+		}
+	}
+
+	fileSaver = async (content: string): Promise<string> => {
+		if (!this._keyBuffer || !this._ivBuffer) {
+			await this.init();
+		}
+		const cipherIv = crypto.createCipheriv('aes-256-cbc', this._keyBuffer!, this._ivBuffer!);
+		return `${cipherIv.update(content, 'utf8', 'base64')}${cipherIv.final('base64')}`;
+	}
+
+	fileOpener = async (content: string): Promise<string> => {
+		if (!this._keyBuffer || !this._ivBuffer) {
+			await this.init();
+		}
+		const decipherIv = crypto.createDecipheriv('aes-256-cbc', this._keyBuffer!, this._ivBuffer!);
```

Comment:
why do we use `-cbc` here but `-gcm` in ADAL class?

Reply:
I can try using `-gcm` for MSAL too now that I got Aes (.NET) to shake hands with 'crypto' in NodeJs.

_Long story short, I had a few challenge with Aes initially (due to encodings, lengths, etc.) and stumbled with 'crypto-js' (that's where `-cbc` came from) during experiments - until in the end I got it working with 'crypto' itself._

Reply:
One more thing, would this be of concern?
https://learn.microsoft.com/en-us/dotnet/standard/security/cross-platform-cryptography#aes-ccm-and-aes-gcm-on-macos

> On macOS, the system libraries don't support AES-CCM or AES-GCM for third-party code, so the [AesCcm](https://learn.microsoft.com/en-us/dotnet/api/system.security.cryptography.aesccm) and [AesGcm](https://learn.microsoft.com/en-us/dotnet/api/system.security.cryptography.aesgcm) classes use OpenSSL for support. Users on macOS need to obtain an appropriate copy of OpenSSL (libcrypto) for these types to function, and it must be in a path that the system would load a library from by default.

Reply:
I noticed `-gcm` (.NET) is unsupported in 'browser' (web-mode), which we intend to support sometime in future.. security wise we're good since it's equivalent. The only diff is authentication tag not available in `gcm` which doesn't give us benefit since we append it to cache (in ADAL) anyways.

![image](https://user-images.githubusercontent.com/13396919/225709464-9abbde91-d8e3-4a85-b561-7f70fe623216.png)

Reply:
How are we even going to have a shared cache file when the tools service is running in a completely different place than the extension itself?

Reply:
What does different place mean? Can it cross OS boundary?

Reply:
See https://code.visualstudio.com/api/extension-guides/web-extensions for details. Relevant parts : 

```
* Importing or requiring other modules is not supported. importScripts is not available as well. As a consequence, the code must be packaged to a single file.
* Node.js globals and libraries such as process, os, setImmediate, path, util, url are not available at runtime. They can, however, be added with tools like webpack. The [webpack configuration](https://code.visualstudio.com/api/extension-guides/web-extensions#webpack-configuration) section explains how this is done.
* [Extension context](https://code.visualstudio.com/api/references/vscode-api#ExtensionContext) locations (ExtensionContext.extensionUri) and storage locations (ExtensionContext.storageUri, globalStorageUri) are also on a virtual file system and need to go through vscode.workspace.fs.
* Creating child processes or running executables is not possible. However, web workers can be created through the [Worker](https://developer.mozilla.org/en-US/docs/Web/API/Worker) API. This is used for running language servers as described in the [Language Server Protocol in web extensions](https://code.visualstudio.com/api/extension-guides/web-extensions#language-server-protocol-in-web-extensions) section.
```

Although I suppose we had web mode running for a while so maybe most of the work is already done - Karl would be the person to talk to about that. Anyways not something to really worry about now, will just be curious to see how we resolve that disconnect. Hopefully this worker API will be running in the same context as where these virtual files are stored.